### ISSUE TITLE
setting $TMPDIR by default

### ISSUE THREAD

**Issue by eszlari** [flatpak/flatpak#3438]
In the documentation, we can read:

> * If an application uses `$TMPDIR` to contain lock files you may want to add a wrapper script that sets it to `$XDG_RUNTIME_DIR/app/$FLATPAK_ID`.

https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

At the moment apps on Flathub set `$TMPDIR` to:

* `$XDG_RUNTIME_DIR/app/$FLATPAK_ID` (current recommendation in the docs)
* `/var/tmp` (which is `$XDG_CACHE_HOME/tmp`, previous recommendation in the docs)
* `$XDG_CACHE_HOME`

Why isn't this set to a sane default value? (Maybe with an option to override it, if this is even needed?)

Also neither `$XDG_RUNTIME_DIR` (cleared on app shutdown) nor ` $XDG_CACHE_HOME` (never cleared), behave like `/tmp` on the host (cleared on reboot: https://systemd.io/TEMPORARY_DIRECTORIES/). Why isn't `/tmp` in the sandbox simply redirected to e.g. `/tmp/flatpak` on the host?

**Comment by TingPing** [comment]
I'd argue applications using `$TMPDIR` for lock files are in the wrong and should be fixed upstream.

**Comment by TingPing** [comment]
> Why isn't /tmp in the sandbox simply redirected to e.g. /tmp/flatpak on the host?

I do think there would be some value in `/tmp` mapping to `/tmp/app/$FLATPAK_ID` or such to get all of the cleanup behavior and such automatically. The fact lockfiles will just work then is a nice bonus.

**Comment by smcv** [comment]
> `$TMPDIR/app`

If you mean `$TMPDIR` on the host, which typically has `drwxrwxrwt` permissions, you can't safely create a subdirectory in it with a hard-coded name like `app`. Either you create it unsafely (and other people can carry out symlink attacks on you), or you create it with `mkdir` and fail if it already exists (and that's an easy denial of service - another user can create the directory and you won't be able to).

X11 manages to get away with this for `/tmp/.X11`, but only by creating it during boot, before any users have had the opportunity to log in.

> `/tmp/flatpak` on the host

This has the same problems.

The historical design of `/tmp` and `/var/tmp` is really not great, but compatibility concerns mean that it isn't going to change. Avoiding them is often the best approach.

**Comment by smcv** [comment]
> to contain lock files

Lock files are a cooperative mutual exclusion mechanism, and appropriate places for lock files depend who you are trying to exclude: other instances of the same app? Other programs running as the same uid? Other programs running as different uids? ...

**Comment by Erick555** [comment]
> Also neither $XDG_RUNTIME_DIR (cleared on app shutdown) nor $XDG_CACHE_HOME (never cleared), behave like /tmp on the host (cleared on reboot: https://systemd.io/TEMPORARY_DIRECTORIES/). Why isn't /tmp in the sandbox simply redirected to e.g. /tmp/flatpak on the host?

I think for sandboxed apps, clearing their temporary dirs on app shutdown is expected feature. I don't see the reason for keeping their contents until shutdown (for host /tmp, system shutdown is real "app shutdown"). `$XDG_CACHE_HOME/tmp` behaves same as `/var/tmp` on most distros (IIRC some bind `/var/tmp` to `/tmp`).

**Comment by Erick555** [comment]
Heh, I just tested that `$XDG_RUNTIME_DIR/app/<appid>` isn't cleared on app shutdown so this whole discussion is misguided 😄 . Some apps like Thunderbird/FF try to clean their temp files on their own.

**Comment by Erick555** [comment]
Nevertheless exporting `TMPDIR=$XDG_RUNTIME_DIR/app/$FLATPAK_ID` and binding `/tmp` to `$XDG_RUNTIME_DIR/app/$FLATPAK_ID`would save many apps from creating wrappers and setting `TMPDIR` themselves as everything will works out of the box then.

@alexlarsson what do you think about above?

**Comment by TingPing** [comment]
The runtime dir has very different behavior than `/tmp`..

**Comment by Erick555** [comment]
@TingPing what behavior? The fact that `/tmp` isn't shared between app instances and can't be used for portals make it broken for many apps so that behavior is unwanted. Other than that both are just tmpfs filesystems.

**Comment by TingPing** [comment]
@Erick555 They just have entirely different purposes, to quote the xdg spec: *" Applications should use this directory for communication and synchronization purposes and should not place larger files in it"*

**Comment by Erick555** [comment]
Where flatpak apps should place larger (or smaller) temporary files then if `/tmp` isn't suitable for it? `$XDG_CACHE_DIR/tmp` is persistent and not cleaned so it may accumulate junk.

Way too many apps on flathub need `TMPDIR` hacks so generic flatpak solution would be helpful here.

> /tmp for example is often automatically cleaned out.

On host it's cleaned out after reboot or after 10 days depending on which happens first. In flatpak it's discarded on app exit. I don't see why this matters for this discussion.

**Comment by TingPing** [comment]
> Where flatpak apps should place larger (or smaller) temporary files then if /tmp isn't suitable for it?

`/var/tmp` on the *host* is correct. Its not kept in memory but its still cleaned up automatically.

> $XDG_CACHE_DIR/tmp is persistent and not cleaned so it may accumulate junk.

Yes `/var/tmp` in the sandbox is bad.

**Comment by Erick555** [comment]
> /var/tmp on the host is correct. Its not kept in memory but its still cleaned up automatically.

But `--filesystem=/var/tmp` is a sandbox hole not much better than --filesystem=/tmp`

**Comment by TingPing** [comment]
I've never suggested adding permissions. I'm just saying that is how it should behave.

One idea is `flatpak` makes custom user+app specific directories on the host and puts a conf file in `/etc/tmpfiles.d/` to configure them to behave similar to `/tmp` and `/var/tmp`.

**Comment by TingPing** [comment]
This might make sense:


`~/.local/share/user-tmpfiles.d/flatpak-apps.conf`:
```
#Type Path                                     Mode User Group Age         Argument
d     %h/.var/app/*/cache/tmp                  -    -    -     10d         -
```

Then `systemd-tmpfiles --user --clean`

**Comment by Erick555** [comment]
I think it should be more like:
```
#Type Path                                     Mode User Group Age Argument
e     %h/.var/app/*/cache/tmp                   -    -    -    10d  -
```
however I prefer `TMPDIR` to be on tmpfs like `/tmp` is. Honestly I'm not convinced that one xdg spec sentence should be decisive argument about how flatpak should work.

**Comment by TingPing** [comment]
Yes a tmpfs /tmp is ideal. No `$XDG_RUNTIME_DIR` isn't where it goes. The latter has nothing to do with accomplishing the former.

**Comment by smcv** [comment]
Neither TMPDIR nor /tmp is guaranteed to be on a tmpfs. TMPDIR on the host system can be set to anything, and /tmp on Debian has traditionally been on the root filesystem (it's straightforward to make it a tmpfs, but that isn't the default).

**Comment by Erick555** [comment]
> No $XDG_RUNTIME_DIR isn't where it goes

but is there a technical argument against it or just authoritative statement? If flatpak can put `/var/tmp` on `$XDG_CACHE_HOME/tmp` then it can put `/tmp` on `$XDG_RUNTIME_DIR/app/$FLATPAK_ID` as well.

@smcv /tmp is nowadays automatically mounted as tmpfs by systemd, does debian override systemd defaults? Anyway I don't think that makes any difference, vast majority of systems have /tmp on tmpfs.

**Comment by gasinvein** [comment]
Can flatpak just create some directory like `/tmp/user-$UID/app/$FLATPAK_ID`, mount it to the sandbox and set `$TMPDIR` inside the sandbox to this directory? This would solve both sharing `$TMPDIR` between instances and opening files inside of it on host.

**Comment by Erick555** [comment]
Creating files in world writable directory is always risky, at least it would need partially random filenames but better to avoid that.

**Comment by gasinvein** [comment]
Well, systemd-tmpfiles can handle creating containing dir (`/tmp/user-$UID` in my example) with proper permissions. And `$TMPDIR`, defaulting to `/tmp`, is generally world-writable anyway.

**Comment by Erick555** [comment]
> And $TMPDIR, defaulting to /tmp, is generally world-writable anyway.

Not in flatpak.

**Comment by james-lawrence** [comment]
what is the hold up that has prevented picking a default value and rolling with it that it requires over 4 years to resolve, and untold hours of developer time discovering and resolving this problem in their application when they try to use flatpak? if its simply the dislike of the /tmp and /tmp/var semantics? if so just go the same route as systemd's PrivateTmp setting for flatpak's runtime; give the applications a private writable directory similar to the runtime dir, if nothing else just assign them to the same directory.

does flatpak need someone to write the probably no more than 10 lines of code to handle this? or are we going to continue on the path of not resolving an unnecessary annoyance within the flatpak runtime environment?
